nav2_ceres_costaware_smoother/src/ceres_costaware_smoother.cpp

```diff
@@ -0,0 +1,156 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <string>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "nav2_ceres_costaware_smoother/ceres_costaware_smoother.hpp"
+#include "nav2_core/exceptions.hpp"
+#include "nav2_util/node_utils.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "nav2_costmap_2d/costmap_filters/filter_values.hpp"
+
+#include "pluginlib/class_loader.hpp"
+#include "pluginlib/class_list_macros.hpp"
+
+using nav2_util::declare_parameter_if_not_declared;
+using nav2_util::geometry_utils::euclidean_distance;
+using namespace nav2_costmap_2d;  // NOLINT
+
+namespace nav2_ceres_costaware_smoother
+{
+
+void CeresCostawareSmoother::configure(
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  std::string name, const std::shared_ptr<tf2_ros::Buffer> & tf,
+  const std::shared_ptr<nav2_costmap_2d::CostmapSubscriber> & costmap_sub,
+  const std::shared_ptr<nav2_costmap_2d::FootprintSubscriber> &)
+{
+  auto node = parent.lock();
+  if (!node) {
+    throw std::runtime_error("Unable to lock node!");
+  }
+
+  costmap_sub_ = costmap_sub;
+  tf_ = tf;
+  plugin_name_ = name;
+  logger_ = node->get_logger();
+
+  _smoother = std::make_unique<nav2_ceres_costaware_smoother::Smoother>();
+  _optimizer_params.get(node.get(), name);
+  _smoother_params.get(node.get(), name);
+  _smoother->initialize(_optimizer_params);
+}
+
+void CeresCostawareSmoother::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up smoother: %s of type"
+    " nav2_ceres_costaware_smoother::CeresCostawareSmoother",
+    plugin_name_.c_str());
+}
+
+void CeresCostawareSmoother::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating smoother: %s of type "
+    "nav2_ceres_costaware_smoother::CeresCostawareSmoother",
+    plugin_name_.c_str());
+}
+
+void CeresCostawareSmoother::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating smoother: %s of type "
+    "nav2_ceres_costaware_smoother::CeresCostawareSmoother",
+    plugin_name_.c_str());
+}
+
+bool CeresCostawareSmoother::smooth(nav_msgs::msg::Path & path, const rclcpp::Duration & max_time)
+{
+  if (path.poses.size() < 2) {
+    return true;
+  }
+
+  auto costmap = costmap_sub_->getCostmap();
+  std::unique_lock<nav2_costmap_2d::Costmap2D::mutex_t> lock(*(costmap->getMutex()));
+
+  // populate smoother input with (x, y, forward/reverse dir)
+  std::vector<Eigen::Vector3d> path_world;
+  path_world.reserve(path.poses.size());
+  Eigen::Vector2d start_dir;
+  Eigen::Vector2d end_dir;
+  for (size_t i = 0; i < path.poses.size(); i++) {
+    auto & pose = path.poses[i].pose;
+    tf2::Quaternion q;
```

Comment:
This can be significantly simplified using tf2's `getYaw()` method. If you give it a quaternion, it will spit out the angle theta of the Z rotation for mobile robot needs. Example: https://github.com/ros-planning/navigation2/blob/fee433ff76b1fe2997043bec16b81568fbead49e/nav2_recoveries/plugins/spin.cpp#L87